### PR TITLE
CAS-16: New grpc interface for Mayastor

### PR DESCRIPTION
**Pull request by yannis218**
Mayastor now implements its own gRPC interface,
so that calls no longer have to go via the agent.

### CODE REVIEW

File: mayastor-test/test_csi.js [hunk 1]
```diff
@@ -111,7 +112,7 @@ function getFsType(mp) {
 }
 
 describe('csi', function() {
-  this.timeout(10000); // for network tests we need long timeouts
+  this.timeout(60000); // for network tests we need long timeouts
```
Comment (jkryl) [inline mayastor-test/test_csi.js:115]: 60s is a bit too much. Why did you change this?

File: mayastor-test/test_rebuild.js [hunk 2]
```diff
@@ -141,6 +141,8 @@ describe('rebuild tests', function() {
     DESTINATION_CHILD: 2,
   };
 
+  this.timeout(60000); // for network tests we need long timeouts
+
```
Comment (jkryl) [inline mayastor-test/test_rebuild.js:145]: the same question here

File: mayastor-test/test_replica.js [hunk 3]
```diff
@@ -60,6 +60,7 @@ function destroyTestDisk(done) {
 }
 
 function createGrpcClient(service) {
+  console.log('Starting client at ', endpoint);
```
Comment (jkryl) [inline mayastor-test/test_replica.js:63]: I suppose this was used for debugging and can be removed now

File: csi/src/server.rs [hunk 4]
```diff
@@ -197,6 +197,7 @@ async fn main() -> Result<(), String> {
     builder.init();
 
     let saddr = format!("{}:{}", addr, port).parse().unwrap();
+    println!("Agent starting service on {}", saddr);
```
[the remark this answers is not in the document]
Comment (yannis218) [reply csi/src/server.rs:200]: Why not?

File: mayastor/src/core/env.rs [hunk 5]
```diff
@@ -110,6 +119,8 @@ pub struct MayastorCliArgs {
 impl Default for MayastorCliArgs {
     fn default() -> Self {
         Self {
+            addr: "None".into(),
```
[the remark this answers is not in the document]
Comment (yannis218) [reply mayastor/src/core/env.rs:122]: It is not an Option; it is a String, for reasons already discussed.  The specific value does not matter, so if you do not like "None", fine, but I do not think that "" would be a good idea, again as already discussed.

File: mayastor/src/core/env.rs [hunk 6]
```diff
@@ -169,6 +180,9 @@ type Result<T, E = EnvError> = std::result::Result<T, E>;
 #[derive(Debug, Clone)]
 pub struct MayastorEnvironment {
     pub config: Option<String>,
+    pub enable_grpc: bool,
+    grpc_addr: String,
```
[the remark this answers is not in the document]
Comment (yannis218) [reply mayastor/src/core/env.rs]: Already an option - you were not looking at the latest diff.

File: mayastor-test/test_cli.js [hunk 7]
```diff
@@ -54,6 +54,8 @@ function runMockServer(rules) {
 }
 
 describe('cli', function() {
+  this.slow(300);
+
```
[the remark this answers is not in the document]
Comment (yannis218) [reply mayastor-test/test_cli.js:58]: Not necessary at all - it only configures the colouring in the console output.  Too many of the tests were being tagged with red for "being slow" (at least on my loaded system) that it became difficult to notice actual test failures.

File: mayastor-test/test_csi.js (hunk 1 above)
Comment (yannis218) [reply mayastor-test/test_csi.js:115]: I am happy to try smaller values, but 10sec on my system fails the test with timeout.  Given that this is basically to guard against tests that hang, what is the harm in using 1min?  If you envisage a test case that has to take less than a particular amount of time to be considered a pass, then that can surely be asserted explicitly by that test.  This is better that having an entire test suite that relies on a performant system.

File: mayastor-test/test_rebuild.js (hunk 2 above)
Comment (yannis218) [reply mayastor-test/test_rebuild.js:145]: See above.

File: mayastor-test/test_replica.js (hunk 3 above)
Comment (yannis218) [reply mayastor-test/test_replica.js:63]: I took out a lot of my debugging prints but not this, as I thought it was not unreasonable for the test to announce the socket for the interface the test suite was about to test.

File: csi/src/client.rs [hunk 8]
```diff
@@ -378,6 +458,63 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                                .index(1),
+                        )
+                        .arg(
+                            Arg::with_name("key")
+                                .help("crypto key to use")
+                                .required(false)
+                                .index(2),
+                        )
+                )
+        )
```
[the remark this answers is not in the document]
Comment (yannis218) [reply csi/src/client.rs:517]: I do not know - these were Jeffry's changes that I "inherited".

File: mayastor-test/test_cli.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (yannis218) [reply mayastor-test/test_cli.js:58]: I am removing this.

File: mayastor-test/test_csi.js (hunk 1 above)
Comment (yannis218) [reply mayastor-test/test_csi.js:115]: I tried reverting this on my new setup (running on the host, no VM) and the tests passed!
(I should have tried doing this in the first place, when I got rid of the VM - it was obvious that everything was running faster)  so I am reverting this.

File: mayastor-test/test_replica.js (hunk 3 above)
Comment (yannis218) [reply mayastor-test/test_replica.js:63]: Particularly as some of the tests now run twice, with different sockets.

Comment (jkryl) [reply mayastor-test/test_replica.js:63]: In general there should be no output from mocha test suite except the green check marks (or red if it fails). That said sometimes the code we are testing logs messages to stderr - nothing we can do about it. And perhaps there are other few exceptions like printing log output of mayastor process or csi agent process (as there is no way to get this output otherwise). But log messages like "Starting something ... ", "Doing this or that" should be avoided as they clutter the output more than necessary. In the long term we won't run the same tests against two endpoints, but there will be just one endpoint. Then how useful is the information that I'm running the tests against some particular endpoint. But we can remove that later when the obsolete code in csi-agent goes away 👍

File: csi/src/client.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkryl) [reply csi/src/client.rs:517]: I seems you inherited the code with its problems too :-) And the problem of this code is that it does not seem to be tested. I'm fine if you just create a ticket addressing the missing tests and tackle it in the next sprint 👍  Or remove this code and push it along with the tests as it should have been done. Whatever you prefer.

File: csi/src/server.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkryl) [reply csi/src/server.rs:200]: why not to use log functions? The same comment for the other println you added below.

File: mayastor-test/test_replica.js (hunk 3 above)
Comment (yannis218) [reply mayastor-test/test_replica.js:63]: I can see the reason you are describing, but when I run the tests I get thousands of lines of output, as it dumps out the std{out,err} of the processes it was running.  If all you got was the grey text with the ticks, I would have hesitated to add this kind of thing.  So OK, I'll remove it again when I get rid of the agent's gRPC and the tests go back to a single socket.

File: csi/src/server.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yannis218) [reply csi/src/server.rs:200]: Oh, as in change the println! to info!
